### DISCUSSION TITLE
Link Time Optimization - compilation and linking on different machines

### DISCUSSION THREAD

**Discussion by Over17** [android/ndk#1750] [General]
Hi!

After we've upgraded to r23b and finally switched the linker to lld (used to be bfd.ld for arm64 builds because issues!), now I'm looking into enabling LTO.

It works on my machine (c), although x86_64 build has increased in size (!) by 1%. (x86 shows best improvement, Arm gains ~1%). I haven't assessed performance yet too.

However, I have a question which is related to our workflow and a feature called "native engine stripping".
- for mobiles, we ship both libunity.so (full) and a static libunityruntime.a (used when engine stripping is enabled) to our users
- the user writes the user's code, builds the game
- when building the game, a special build step creates a map of functions/modules used
- unused modules are stripped from the .a file (for example, if you don't use 2D physics, it will be gone)
- final .so is linked on the user machine using the exact same NDK version we used on our build servers (r23b).

APK-size-wise, this optimization gives huge gains.

Can LTO be enabled in such a case? Does it matter if the compilation and linking happens on different machines (and maybe even different OS's)? 

Thanks!

**Comment by DanAlbert** [comment]
Afaik that should work fine as long as the user is using the same version of LLVM (should be compatible within any NDK major release) that was used to build libunityruntime.a. If it isn't, that sounds like a bug.

I don't think it'd be a good thing for most library vendors since it could make getting agreement on the compiler version very difficult for the app developer, but Unity seems like a reasonable exception to me, since it already provides pretty clear guidance in which NDKs to use with each version of the engine.

As an aside, r23's support window ended recently. If you're having issues migrating to r25, please let us know so we can help :)

**Reply by pirama-arumuga-nainar** [reply]
+1 - this should be fine with the caveats @DanAlbert mentioned.

**Reply by Over17** [reply]
Thank you - will see how far I can get with it!

Re r25... it's a complicated topic; if it takes another month of comparing perf results and finding LLVM workarounds, then I'd rather refrain from upgrading. However the native M1 binaries are a big plus.

**Reply by DanAlbert** [reply]
Btw, if you're able, we'd _love_ to run your benchmarks as part of our own testing. Writing our own benchmark suite isn't something we've been funded for, but Unity's has certainly found interesting results over the years. Feel free to email me directly (address listed on my profile page) if that's something you're interested in. If you need a way to sell that to your boss, remind them how much easier NDK upgrades would be if we've already run your benchmarks :)

**Reply by Over17** [reply]
I'll see what we can do... our performance tests are built in into the code so it would be hard to extract them to use as standalone.

**Comment by Over17** [comment]
Fun fact:
.so being smaller with LTO, but the compressed size is larger (!!!)
Wondering if Brotli or whatever is used to compress OTA packages produces a different result.

![LTO_sizes](https://user-images.githubusercontent.com/6140634/185105794-7d757c54-6639-4811-b6ee-b8b06eee0af6.PNG)

**Comment by Over17** [comment]
Any ideas why I have an LTO'ed build crashing early during init?

> 08-17 15:19:19.467 E/CRASH   (27736): *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
> 08-17 15:19:19.467 E/CRASH   (27736): Version '2023.1.0a6 (d9d74c5026b0)', Build type 'Release', Scripting Backend 'il2cpp', CPU 'arm64-v8a'
> 08-17 15:19:19.467 E/CRASH   (27736): Build fingerprint: 'samsung/r0sxeea/r0s:12/SP1A.210812.016/S901BXXU2AVG6:user/release-keys'
> 08-17 15:19:19.467 E/CRASH   (27736): Revision: '28'
> 08-17 15:19:19.467 E/CRASH   (27736): ABI: 'arm64'
> 08-17 15:19:19.467 I/GestureDetector( 1228): handleMessage TAP
> 08-17 15:19:19.467 E/CRASH   (27736): Timestamp: 2022-08-17 15:19:19.467549536+0300
> 08-17 15:19:19.467 E/CRASH   (27736): pid: 27736, tid: 27767, name: UnityMain  >>> com.unity3d.torture <<<
> 08-17 15:19:19.467 E/CRASH   (27736): uid: 10341
> 08-17 15:19:19.467 E/CRASH   (27736): signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr --------
> 08-17 15:19:19.467 E/CRASH   (27736):     x0  0000007353d99100  x1  00000073526ea85f  x2  0000000000000004  x3  0000000000000010
> 08-17 15:19:19.467 E/CRASH   (27736):     x4  0000000000000000  x5  00000073527b0836  x6  00000073200037a4  x7  6f65726574535f79
> 08-17 15:19:19.467 E/CRASH   (27736):     x8  0000007353d99000  x9  000000000000001f  x10 0000000000000020  x11 0000000000000000
> 08-17 15:19:19.467 E/CRASH   (27736):     x12 0000000000000001  x13 0000000000000000  x14 000000000000026d  x15 0000000000000000
> 08-17 15:19:19.467 E/CRASH   (27736):     x16 0000007353ccbd88  x17 00000076d65ae000  x18 00000073524b6000  x19 0000007352827568
> 08-17 15:19:19.467 E/CRASH   (27736):     x20 00000073526ea84b  x21 0000007320003790  x22 0000000000000014  x23 0000007353d99000
> 08-17 15:19:19.467 E/CRASH   (27736):     x24 00000000811c9dc5  x25 00000000bfffffff  x26 ffffffffffffffff  x27 0000000000000000
> 08-17 15:19:19.467 E/CRASH   (27736):     x28 00000073001cc510  x29 00000073001cc290
> 08-17 15:19:19.467 E/CRASH   (27736):     lr  00000073532567cc  sp  00000073b4b5d8c0  pc  00000073532567f4  pst 0000000060001000
> 

ACCERR without an address is weird, I compared binary sections etc but they are identical. Could it be compiler/linker flags conflicting? Or some UB which works fine until you LTO it? Suspecting static initialization order issues...

**Reply by enh-google** [reply]
worth disabling your crash reporter and just letting the system's one have a go, see if that gets different/better results? (not least because -- if not -- that's a bug in its own right that we should look at :-) )

**Reply by Over17** [reply]
@enh-google FYI yes the default crash handler was able to publish the address etc. so there's a bug in our crash handler. we were also able to find and fix a UB in our code which was causing the crash.

**Reply by jmgao** [reply]
@Over17 

As a sidenote, is there any chance that you could call through to the default system-registered signal handler after you're done in yours? Unity replacing the default signal handler makes it so that crashes in games don't get funneled through pipelines that look at tombstones or modify libdebuggerd_handler, so neither Google nor vendors (Oculus in my case) will notice when they make a change that increases the crash rates for unity apps.

**Reply by Over17** [reply]
@jmgao this is being fixed and has a chance to make it into Unity 2023.1.

**Reply by jmgao** [reply]
> @jmgao this is being fixed and has a chance to make it into Unity 2023.1.

Awesome, thanks!

**Comment by Over17** [comment]
@DanAlbert finally our LTO got to a CI, and here's a crash:

> [16:42:21] ##### ExitCode
> [16:42:21] 254
> [16:42:21] ##### Output
> [16:42:21] clang++: warning: argument unused during compilation: '-mbranch-protection=standard' [-Wunused-command-line-argument]
> [16:42:21] LLVM ERROR: out of memory
> [16:42:21] Allocation failed
> [16:42:21] PLEASE submit a bug report to https://github.com/android-ndk/ndk/issues and include the crash backtrace.
> [16:42:21] Stack dump:
> [16:42:21] 0. Program arguments: "External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld" @/tmp/response-ee5862.txt
> [16:42:21] 1. Running pass 'Function Pass Manager' on module 'ld-temp.o'.
> [16:42:21] 2. Running pass 'Machine Instruction Scheduler' on function '@_ZN11NoiseModule17CalculateNoiseJobILi2ELb1ELb1EEEvP12NoiseJobDataj'
> [16:42:21] Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamEi+0x18)[0x25bcfa8]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm3sys17RunSignalHandlersEv+0x50)[0x25bce30]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld[0x25bd47a]
> [16:42:21] /lib/x86_64-linux-gnu/libpthread.so.0(+0x12980)[0x7f6922593980]
> [16:42:21] /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xc7)[0x7f69214f6fb7]
> [16:42:21] /lib/x86_64-linux-gnu/libc.so.6(abort+0x141)[0x7f69214f8921]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm22report_bad_alloc_errorEPKcb+0x7e)[0x258caee]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm25llvm_unreachable_internalEPKcS1_j+0x0)[0x2593f40]
> [16:42:21] /home/bokken/buildslave/unity/build/External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/../lib64/libc++.so.1(_Znwm+0x2b)[0x7f6921b3cbcb]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld[0x2024a87]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld[0x202354e]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm17ScheduleDAGMILive8scheduleEv+0x3a)[0x1dd54da]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld[0x1e427ce]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld[0x1e424bb]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm19MachineFunctionPass13runOnFunctionERNS_8FunctionE+0x9a)[0x1ff272a]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm13FPPassManager13runOnFunctionERNS_8FunctionE+0x2e7)[0x1a08fe7]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm13FPPassManager11runOnModuleERNS_6ModuleE+0x2f)[0x1a08cdf]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE+0x2a4)[0x2280d44]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld[0x2b89092]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm3lto7backendERKNS0_6ConfigENSt3__18functionIFNS4_10unique_ptrINS0_18NativeObjectStreamENS4_14default_deleteIS7_EEEEjEEEjNS6_INS_6ModuleENS8_ISD_EEEERNS_18ModuleSummaryIndexE+0x9d)[0x2b78e3d]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm3lto3LTO13runRegularLTOENSt3__18functionIFNS2_10unique_ptrINS0_18NativeObjectStreamENS2_14default_deleteIS5_EEEEjEEE+0x448)[0x2b77f18]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN4llvm3lto3LTO3runENSt3__18functionIFNS2_10unique_ptrINS0_18NativeObjectStreamENS2_14default_deleteIS5_EEEEjEEENS3_IFSA_jNS_9StringRefEEEE+0x1de)[0x26085de]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN3lld3elf15BitcodeCom
> piler7compileEv+0x160)[0x265be00]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN3lld3elf12LinkerDriver19compileBitcodeFilesIN4llvm6object7ELFTypeILNS3_7support10endiannessE1ELb1EEEEEvv+0x7c)[0x26547cc]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN3lld3elf12LinkerDriver4linkIN4llvm6object7ELFTypeILNS3_7support10endiannessE1ELb1EEEEEvRNS3_3opt12InputArgListE+0x5e1)[0x2647f21]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN3lld3elf12LinkerDriver10linkerMainEN4llvm8ArrayRefIPKcEE+0x4f1)[0x2641151]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(_ZN3lld3elf4linkEN4llvm8ArrayRefIPKcEEbRNS1_11raw_ostreamES7_+0x139)[0x258c529]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld[0x18fcf41]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld(main+0x69)[0x18fc979]
> [16:42:21] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f69214d9bf7]
> [16:42:21] External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld[0x340b591]
> [16:42:21] clang++: error: unable to execute command: Killed
> [16:42:21] clang++: error: linker command failed due to signal (use -v to see invocation)
> [16:42:21] [5139/5152 564s] Link_Android_arm64 artifacts/AndroidPlayer/Android_arm64_nondev_i_r/libunity.so
> [16:42:21] ##### CommandLine
> [16:42:21] "External/NonRedistributable/Android ndk/builds/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++" @"artifacts/rsp/3349697964252950941.rsp"

happy to share the relevant contents of the response file if needed.

NDK: r23b

build agent: ubuntu VM, 16Gb RAM, 8 cores.

**Reply by stephenhines** [reply]
> [16:42:21] LLVM ERROR: out of memory
> [16:42:21] Allocation failed

This looks like a traditional OOM error, so I'm not sure that there's something actionable for us (other than trying to make LLVM use less memory). Depending on the size of what you're LTOing, 16Gb may not be enough, or you might just need to have your build system run fewer concurrent memory-heavy operations like link steps. Linking is already very memory intensive, and LTO makes this worse by pushing more optimizations + needing to have all of the sources/libraries available in memory to finalize the code.

**Reply by Over17** [reply]
Fair enough thank you Stephen.

Is ThinLTO stable enough to be used? Checked it out and it seems to work - less memory, shorter link time; however the final binary size is +500k on 27meg. I pass -Os to the compiler (except for perf critical pieces) and the linker - is it supposed to configure what LTO optimizes for?

Looking at ld.lld response file (not the one we pass to clang), "-plugin-opt=O2" is there, so maybe -Os is lost somewhere or not supposed to work at all? +500k is quite an upsetting result for -Os.

**Reply by stephenhines** [reply]
> Is ThinLTO stable enough to be used? Checked it out and it seems to work - less memory, shorter link time; however the final binary size is +500k on 27meg. I pass -Os to the compiler (except for perf critical pieces) and the linker - is it supposed to configure what LTO optimizes for?

We use ThinLTO extensively for Android's most important libraries (https://cs.android.com/search?q=file:Android.bp%20%22thin:%22), and are even trying to use ThinLTO globally in a size-reducing form (https://android-review.googlesource.com/q/topic:global-thinlto, but we are still fixing random issues re: non-exported symbols disappearing, which is why this has been landed/reverted a few times). You might have different experiences, but please file bugs if you encounter any trouble.

> Looking at ld.lld response file (not the one we pass to clang), "-plugin-opt=O2" is there, so maybe -Os is lost somewhere or not supposed to work at all? +500k is quite an upsetting result for -Os.

In general, I think that the recommendation has been to use `-Oz` with Clang, as `-Os` isn't very aggressive about code size, and still prefers performance optimizations in most cases (see https://android.googlesource.com/platform/ndk/+/master/docs/BuildSystemMaintainers.md#controlling-binary-size for a bit more info about this). We're using `-Wl,--lto-O0` (https://cs.android.com/android/platform/superproject/+/master:build/soong/cc/lto.go;l=100) to keep ThinLTO more focused on reducing code size/duplication than performance optimizations on non-perf-critical libraries. I'm not surprised that the default optimization-heavy path would still lead to increased code size, since you really do need to pass `-Oz` to Clang for size savings similar to GCC's `-Os`.
